rage/vector.hpp

```diff
@@ -27,22 +27,133 @@ namespace rage
 		T data[3];
 		struct { T x, y, z; };
 
-		vector3(T x, T y, T z) :
-			x(x),
+		vector3(T x, T y, T z) noexcept : x(x),
 			y(y),
 			z(z)
 		{
 		}
 
-		vector3() :
-			x(),
+		vector3() noexcept : x(),
 			y(),
 			z()
 		{
 		}
+
+		vector3 operator+(const vector3 &rhs) const
+		{
+			return vector3(x + rhs.x, y + rhs.y, z + rhs.z);
+		}
+
+		vector3 operator-(const vector3 &rhs) const
+		{
+			return vector3(x - rhs.x, y - rhs.y, z - rhs.z);
+		}
+
+		vector3 operator*(const vector3 &rhs) const
+		{
+			return vector3(x * rhs.x, y * rhs.y, z * rhs.z);
+		}
+
+		vector3 operator/(const vector3 &rhs) const
+		{
+			return vector3(x / rhs.x, y / rhs.y, z / rhs.z);
+		}
+
+		vector3 operator+(const float &rhs) const
+		{
+			return vector3(x + rhs, y + rhs, z + rhs);
+		}
+
+		vector3 operator-(const float &rhs) const
+		{
+			return vector3(x - rhs, y - rhs, z - rhs);
+		}
+
+		vector3 operator*(const float &rhs) const
+		{
+			return vector3(x * rhs, y * rhs, z * rhs);
+		}
+
+		vector3 operator/(const float &rhs) const
+		{
+			return vector3(x / rhs, y / rhs, z / rhs);
+		}
+
+		bool operator==(const vector3 &rhs) const
+		{
+			return x == rhs.x && y == rhs.y && z == rhs.z;
+		}
+
+		vector3 &operator+=(const vector3 &rhs)
+		{
+			return *this = *this + rhs;
+		}
+
+		vector3 &operator-=(const vector3 &rhs)
+		{
+			return *this = *this - rhs;
+		}
+
+		vector3 &operator*=(const vector3 &rhs)
+		{
+			return *this = *this * rhs;
+		}
+
+		vector3 &operator/=(const vector3 &rhs)
+		{
+			return *this = *this / rhs;
+		}
+
+		vector3 &operator+=(const float &rhs)
+		{
+			return *this = *this + rhs;
+		}
+
+		vector3 &operator-=(const float &rhs)
+		{
+			return *this = *this - rhs;
+		}
+
+		vector3 &operator*=(const float &rhs)
+		{
+			return *this = *this * rhs;
+		}
+
+		vector3 &operator/=(const float &rhs)
+		{
+			return *this = *this / rhs;
+		}
+
+		float length() const
+		{
+			return sqrt(x * x + y * y + z * z);
+		}
+
+		// This doesn't modify the vector inline.
+		vector3 normalize() const
+		{
+			return *this * (1 / length());
+		}
+
+		// This doesn't modify the vector inline.
+		vector3 multiply(const vector3 other) const
+		{
+			return vector3(x * other.x, y * other.y, z * other.z);
+		}
+
+		// This doesn't modify the vector inline.
+		vector3 invert() const
+		{
+			return *this * -1;
+		}
+
+		float distance(const vector3 &rhs) const
+		{
+			return (*this - rhs).length();
+		}
```

Comment:
Some nice to have functions would be Zero, that just zeros out the vector. This would be useful instead of instantiating a new vector with 0,0,0 then setting it. 

[Cross Product](https://en.wikipedia.org/wiki/Cross_product). 

```
vector3 crossProduct(const vector3& rhs)
{
    return vector3(y * rhs.z - z * rhs.y, z * rhs.x - x * rhs.z, x * rhs.y - y * rhs.x);
}
```